(PUP-920) Add additional information to the report

melissa opened the pull request.

joshcooper commented on lib/puppet/configurer.rb:
```diff
@@ -102,32 +102,44 @@ def retrieve_catalog(query_options)
+        Puppet::GettextConfig.reset_text_domain('agent')
+        Puppet::ModuleTranslations.load_from_vardir(Puppet[:vardir])
+      end
+      options[:report].add_times(:plugin_sync, plugin_sync_time) if options[:report]
```
Looks good. One nit is that the agent may make multiple passes if the environment doesn't match. So maybe we need to accumulate `plugin_sync`, `fact_generation`, and `node_retrieval` times? e.g `options[:report].add_times(:plugin_sync, pluginsync_time, true)` and add an optional arg to `Report#add_times` to either overwrite or accumulate the metric?

melissa replied:
I was actually just thinking about trying to figure out what would happen in that case. Yes, that sounds like a good solution here. I'll look into adding that in.

melissa replied:
This is done, but I modified the `add_times` method to automatically accrue metrics unless otherwise specified. I think we don't want to accidentally overwrite time data just in case we guess wrong  about which methods are called multiple times.

puppetcla commented:
CLA signed by all contributors.

joshcooper commented on lib/puppet/configurer.rb:
```diff
@@ -258,38 +270,40 @@ def run_internal(options)
               :fail_on_404 => true)
+          end
+          options[:report].add_times(:node_retrieval, node_retr_time, true)
+          raise unless node
```
I think this raise isn't necessary because we pass `fail_on_404 => true` above (so the `find` method will raise instead of returning `nil`). If that happens we fall through to the rescue below, print a warning, and keep the agent run going.

joshcooper commented on lib/puppet/transaction/report.rb:
```diff
@@ -151,8 +151,13 @@ def <<(msg)
-    @external_times[name] = value
+  def add_times(name, value, accumulate = false)
+    if @external_times[name] && accumulate
+      puts 'adding times'
```
extra debugging

joshcooper commented on lib/puppet/transaction/report.rb:
```diff
@@ -151,8 +151,13 @@ def <<(msg)
+  def add_times(name, value, accumulate = false)
+    if @external_times[name] && accumulate
+      puts 'adding times'
+      @external_times[name] += value
```
Does `+=` work if the value hasn't been set yet?

joshcooper commented:
One cool thing is that these metrics are automatically reported with the `--summarize` option:

```
$ bx puppet agent -t --summarize
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Loading facts
Info: Caching catalog for localhost
Info: Applying configuration version '1519668896'
Notice: Applied catalog in 0.01 seconds
Changes:
Events:
Resources:
            Total: 7
Time:
         Schedule: 0.00
   Transaction evaluation: 0.01
   Convert catalog: 0.03
   Node retrieval: 0.11
   Config retrieval: 0.14
      Plugin sync: 0.40
         Last run: 1519668896
   Fact generation: 2.24
            Total: 2.93
       Filebucket: 0.00
Version:
           Config: 1519668896
           Puppet: 5.5.0
```

But it looks like the `Puppet::Transaction::Report#summary` method incorrectly sorts `Total` (it's supposed to be last, but it shows up before `Filebucket`).

Also I think we want to change `config_retrieval` and `transaction_evaluation` metrics to also pass the `true` accumulator arg, as the agent will request a new catalog if the catalog it initially receives is for the wrong environment.

melissa commented:
```
Changes:
Events:
Resources:
            Total: 305
Time:
       Filebucket: 0.00
          Package: 0.00
         Schedule: 0.00
             Cron: 0.01
             File: 0.08
          Service: 0.09
   Node retrieval: 0.29
   Convert catalog: 0.30
   Profile manager: 0.33
      Plugin sync: 1.16
             Exec: 1.62
         Last run: 1519845037
            Total: 17.41
   Config retrieval: 4.11
   Transaction evaluation: 4.49
   Fact generation: 4.94
        Pe anchor: 0.00
Version:
           Config: 1519845029
           Puppet: 5.5.0
```

I'm still seeing issues with the ordering of reported times, so I'll continue digging into that. It might be worth opening a new ticket for though

melissa commented:
Also tests. @joshcooper what kinds of tests are you hoping to see for these changes?

melissa commented on lib/puppet/configurer.rb:
```diff
@@ -258,38 +270,40 @@ def run_internal(options)
               :fail_on_404 => true)
+          end
+          options[:report].add_times(:node_retrieval, node_retr_time)
+          raise if node.nil?
```
@joshcooper @jhelwig pointed out that we could still get nil if we get anything other than a 404 or 200 response https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/rest.rb#L190 so I added this line back

joshcooper replied:
Ok sounds good.

joshcooper commented on lib/puppet/configurer.rb:
```diff
@@ -102,22 +102,31 @@ def retrieve_catalog(query_options)
+      catalog.write_class_file
+      catalog.write_resource_file
+    end
+    options[:report].add_times(:convert_catalog, catalog_converstion_time) if options[:report]
```
Since the `thinmark` logic is repeated, what about creating a helper method:

```ruby
def timed_block(report, metric, &block) do
  result = nil 
  time = thinmark do
    result = yield
  end
  report.add_times(metric, time) if report
  result
end
```

And then calling it from the various places:

```ruby
catalog = timed_block(options[:report], :convert_catalog) do
  catalog = result.to_ral
  catalog.write_resource_file
  catalog.finalize
  catalog.retrieval_duration = duration
  catalog.write_class_file
  catalog.write_resource_file
  catalog
end

melissa replied:
There are only two locations where this would be needed, so I'm not sure it's worth it quite yet. I can still do this if you think it's worth it though

joshcooper replied:
ah no worries, that's fine

joshcooper commented:
Looks good to me. The metrics add up correctly which is cool. Did you want to fix the `Total` so that line shows up last?

melissa commented:
I couldn't find an easy place to add tests for the new times for `add_facts` or `run_internal`, but those were the only two I didn't add tests for

melissa commented:
I'm still trying to figure out how to get total to be at the end. It's surprisingly not obvious, at least not to me. That, and @jhelwig pointed out that total may not be entirely accurate. Some of these times may be subset of other reported times. That, and I was thinking, there may be some times unaccounted for. It would be nice if we could wrap the whole puppet run in a time block and get total from that...

joshcooper commented:
You could add an integration test under `spec/integration/apply` that programmatically invokes the equivalent of `puppet apply -e "...something..." --summarize` and then verify the report contains the expected values:

```yaml
metrics:
  time:
    name: time
    label: Time
    values:
    - - notify
      - Notify
      - 0.000854
    - - schedule
      - Schedule
      - 0.000249
    - - filebucket
      - Filebucket
      - 4.2e-05
    - - config_retrieval
      - Config retrieval
      - 0.069442
    - - total
      - Total
      - 0.070587
```

8675309 commented:
Possibly add a test to ensure that the total value is being summed correctly

melissa commented on spec/unit/transaction/report_spec.rb:
```diff
@@ -357,6 +357,36 @@ def add_statuses(count, type = :file)
+
+        @report.finalize_report
+
+        expect(metric(:time, "total")).to eq(18)
```
@8675309 total time test added

melissa commented on lib/puppet/transaction/report.rb:
```diff
@@ -345,9 +349,9 @@ def summary
+        if a == "total"
           1
-        elsif b == :total
+        elsif b == "total"
```
@joshcooper I figured out why total wasn't being presented last in the summary

melissa replied:
These are stringified in `raw_summary`

joshcooper replied:
It'd be good to freeze those strings so we don't create extra object in memory

melissa commented on spec/unit/transaction/report_spec.rb:
```diff
@@ -498,6 +528,10 @@ def add_statuses(count, type = :file)
     end
+
+    it 'should sort total at the very end of the time metrics' do
+      expect(@report.summary).to match(/Total: \d+\.\d{2}\\nVersion:/)
```
This test currently isn't cooperating, so I'm just fighting with white space expectations to get it passing

joshcooper replied:
Maybe need to specify multiline match `/.../m`?

melissa commented:
@joshcooper my gut says an integration test is too heavy weight for this PR, but I can look into adding it if you think it's a good idea

melissa commented:
```
$ sudo bundle exec ./bin/puppet agent -t --summarize
Info: Using configured environment 'workstation_production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Warning: Facter: Fact resolution fact='profiles', resolution='<anonymous>' resolved to an invalid value: Expected 2017-09-08T17:03:53+00:00 to be one of [Integer, Float, TrueClass, FalseClass, NilClass, String, Array, Hash], but was DateTime
Warning: Facter: Fact resolution fact='puppet_classes', resolution='<anonymous>' resolved to an invalid value: Expected all to be one of [Integer, Float, TrueClass, FalseClass, NilClass, String, Array, Hash], but was Symbol
Info: Caching catalog for melissa-c02mq17dfd57
Info: Applying configuration version '1520294824'
Notice: /Stage[main]/Puppet_enterprise::Mcollective::Server/File[/etc/puppetlabs/mcollective/server.cfg]/content: [diff redacted]
Notice: /Stage[main]/Puppet_enterprise::Mcollective::Server/File[/etc/puppetlabs/mcollective/server.cfg]/content: changed [redacted] to [redacted]
Info: /Stage[main]/Puppet_enterprise::Mcollective::Server/File[/etc/puppetlabs/mcollective/server.cfg]: Scheduling refresh of Service[mcollective]
Warning: Cannot read file /var/db/com.apple.xpc.launchd/disabled.plist; Puppet is skipping it.\nDetails: Execution of '/usr/bin/plutil -convert xml1 -o /dev/stdout /var/db/com.apple.xpc.launchd/disabled.plist' returned 1: /var/db/com.apple.xpc.launchd/disabled.plist: Property List error: Cannot parse a NULL or zero-length data / JSON error: No value.
Error: /Stage[main]/Puppet_enterprise::Mcollective::Service/Service[mcollective]: Failed to call refresh: undefined method `[]=' for nil:NilClass
Error: /Stage[main]/Puppet_enterprise::Mcollective::Service/Service[mcollective]: undefined method `[]=' for nil:NilClass
Info: Stage[main]: Unscheduling all events on Stage[main]
Notice: Applied catalog in 6.81 seconds
Changes:
            Total: 1
Events:
          Success: 1
            Total: 1
Resources:
          Changed: 1
   Failed to restart: 1
      Out of sync: 1
            Total: 331
Time:
         Schedule: 0.00
          Package: 0.00
   Property list key: 0.00
          Printer: 0.01
             Cron: 0.01
             File: 0.07
          Service: 0.09
   Node retrieval: 0.24
   Profile manager: 0.28
   Convert catalog: 0.36
      Plugin sync: 0.95
             Exec: 1.62
         Last run: 1520294835
   Config retrieval: 4.17
   Fact generation: 4.97
       Filebucket: 0.00
   Transaction evaluation: 6.71
        Pe anchor: 0.00
            Total: 19.49
Version:
           Config: 1520294824
           Puppet: 5.5.0
```

However, this doesn't address @jhelwig 's concern with the fact that total is now likely no longer accurate (since some of the time metrics we are reporting above are subsets of other time metrics reported above). I created https://tickets.puppetlabs.com/browse/PUP-8511 to deal with that, since it's a bigger issue.

melissa commented on lib/puppet/transaction/report.rb:
```diff
@@ -134,6 +134,8 @@ class Puppet::Transaction::Report
   #
   attr_accessor :transaction_completed
 
+  TOTAL = "total".freeze
```
@joshcooper this is what I ended up with. Is this too heavy handed?

melissa commented on spec/unit/transaction/report_spec.rb:
```diff
@@ -498,6 +528,17 @@ def add_statuses(count, type = :file)
+   Transaction evaluation: \d+.\d{2}
+         Last run: \d+
+            Total: \d+.\d{2}
+Version:
```
I'm a little worried this is too specific to the tests, but hey, it passes!

melissa commented:
![screen shot 2018-03-06 at 11 05 58 am](https://user-images.githubusercontent.com/4473089/37052538-65e25dbc-212e-11e8-8d9f-41f3b5163442.png)
AppVeyor has an interesting failure with this change, but only for windows with Ruby21-x64. It looks like we aren't even getting a total time metric, which is strange to me

ericldelaney commented on lib/puppet/configurer.rb:
```diff
@@ -102,22 +102,31 @@ def retrieve_catalog(query_options)
+  def convert_catalog(result, duration, options = {})
+    catalog = nil
+
+    catalog_converstion_time = thinmark do
```
converstion -> conversion

kris-bosland commented:
Hi @melissa, with the 5.5.x branch being cut, please retarget that branch.

melissa commented:
@joshcooper I'm a little stumped with these tests. Windows is periodically failing because it's not collecting any time stats at all. Because there are no time stats, it's not calculating the total time. Should I update this test to not run on windows, or are there different changes I should make?

joshcooper commented:
Ruby 2.1 implements [`Benchmark.realtime` using `Time.now`](https://github.com/ruby/ruby/blob/ruby_2_1/lib/benchmark.rb#L292-L296), but `Time.now` doesn't have good resolution on Windows:

```
irb(main):016:0> (Time.now - Time.now) == 0
=> true
```

Since it's zero the report summary logic strips out the metric.

In Ruby 2.4, it looks like [`Benchmark.realtime`](https://github.com/ruby/ruby/blob/ruby_2_4/lib/benchmark.rb#L306-L310) is implemented using a better timer:

```
irb(main):017:0> Process.clock_gettime(Process::CLOCK_MONOTONIC) - Process.clock_gettime(Process::CLOCK_MONOTONIC)
=> -1.700012944638729e-06
```

So this makes me wonder, why do we strip out metrics whose value are 0? Maybe the right thing is to not do that?

joshcooper commented:
FYI @scottyw, @ploubser we're adding some timing metrics to the report to give a more complete picture of where the agent spends its time. I'm not sure if you have thoughts about how to surface this in the console or puppet reports?

ploubser commented:
@joshcooper very cool. I can see us wanting to add this information to the metrics tab in reports. I'll bring it up with UX.

joshcooper commented:
👍 

```
Notice: Applied catalog in 9.28 seconds
Changes:
            Total: 1
Events:
          Success: 1
            Total: 1
Resources:
        Restarted: 1
          Changed: 1
      Out of sync: 1
            Total: 331
Time:
        Pe anchor: 0.00
       Filebucket: 0.00
         Schedule: 0.00
          Package: 0.00
   Property list key: 0.00
          Printer: 0.01
             Cron: 0.02
          Service: 0.06
             File: 0.09
   Convert catalog: 0.29
   Profile manager: 0.31
   Node retrieval: 0.83
      Plugin sync: 1.04
             Exec: 1.23
         Last run: 1520982416
   Fact generation: 7.35
   Config retrieval: 9.07
   Transaction evaluation: 9.22
   Catalog application: 9.28
            Total: 38.81
Version:
           Config: 1520982403
           Puppet: 5.5.0
```
